Type parameters checks are inconsistent and incomplete

The way we currently check type parameters is inconsistent and error prone. The code below compiles and it shouldn't.

The `resolve` phase currently ignores these checks and leaves them up to `typeck`, `collect` and `check`, which as demonstrated by bugs like this one and #20302 makes this check unsound. 

``` rust
fn is_copy<T: ::std::marker<i32>::Copy>() {}

fn main() {}
```

Marking as E-needstest.

```
error[E0109]: type parameters are not allowed on this type
 --> test.rs:1:29
  |
1 | fn is_copy<T: ::std::marker<i32>::Copy>() {}
  |                             ^^^ type parameter not allowed

error: aborting due to previous error(s)
```

--- a/src/test/compile-fail/issue-22706.rs
+++ b/src/test/compile-fail/issue-22706.rs
@@ -1,0 +1,13 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn is_copy<T: ::std::marker<i32>::Copy>() {}
+//~^ ERROR type parameters are not allowed on this type [E0109]
+fn main() {}

